File: python/그래프이론/정확한 순위.py
INF = 1e9
n, m = map(int, input().split())

graph = [[INF]*(n+1) for _ in range(n+1)]

#자기 자신은 항상 도달이 가능하다.
for i in range(n+1):
    graph[i][i] = 1

for _ in range(m):
    a, b = map(int, input().split())
    graph[a][b] = 1


for k in range(1,n+1):
    for a in range(1,n+1):
        for b in range(1,n+1):
            graph[a][b] = min(graph[a][b], graph[a][k]+graph[k][b])

result = 0
for a in range(1, n+1):
    count = 0
    for b in range(1, n+1):
        if graph[a][b] != INF or graph[b][a] != INF:
            count += 1
    if count == n:
        result += 1
print(result)

File: JavaScript/최단 경로/기출문제/정확한 순위.js
const n = 6;
const m = 6;
const input = [
  [1, 5],
  [3, 4],
  [4, 2],
  [4, 6],
  [5, 2],
  [5, 4],
];

const graph = Array.from({ length: n + 1 }, () => Array(n + 1).fill(Infinity));

for (let a = 1; a < n + 1; a++) {
  graph[a][a] = 0;
}

for (const [a, b] of input) {
  graph[a][b] = 1;
}

for (let k = 1; k < n + 1; k++) {
  for (let a = 1; a < n + 1; a++) {
    for (let b = 1; b < n + 1; b++) {
      graph[a][b] = Math.min(graph[a][b], graph[a][k] + graph[k][b]);
    }
  }
}

console.log(graph);

let result = 0;
for (let i = 1; i < n + 1; i++) {
  let count = 0;
  for (let j = 1; j < n + 1; j++) {
    if (graph[i][j] !== Infinity || graph[j][i] !== Infinity) {
      count++;
    }
  }
  if (count === n) {
    result++;
  }
}

console.log(result);
